pkg/controller/compute/scheduler/scheduler.go

```diff
@@ -0,0 +1,158 @@
+)
+
+const (
+	controllerName = "scheduler.compute.crossplane.io"
+
+	errorUnschedulable = "Unschedulable"
+)
+
+var (
+	ctx           = context.Background()
+	resultDone    = reconcile.Result{}
+	resultRequeue = reconcile.Result{Requeue: true}
+)
+
+// Add creates a new Instance Controller and adds it to the Manager with default RBAC.
+// The Manager will set fields on the Controller and Start it when the Manager is Started.
+func Add(mgr manager.Manager) error {
+	return add(mgr, newReconciler(mgr))
+}
+
+// Reconciler reconciles a Instance object
+type Reconciler struct {
+	client.Client
+	scheme           *runtime.Scheme
+	recorder         record.EventRecorder
+	lastClusterIndex uint64
+
+	schedule func(*computev1alpha1.Workload) (reconcile.Result, error)
+}
+
+// newReconciler returns a new reconcile.Reconciler
+func newReconciler(mgr manager.Manager) reconcile.Reconciler {
+	r := &Reconciler{
+		Client:   mgr.GetClient(),
+		scheme:   mgr.GetScheme(),
+		recorder: mgr.GetRecorder(controllerName),
+	}
+	r.schedule = r._schedule
+	return r
+}
+
+// CreatePredicate accepts Workload instances with set `Status.Cluster` reference value
+func CreatePredicate(event event.CreateEvent) bool {
+	wl, ok := event.Object.(*computev1alpha1.Workload)
+	if !ok {
+		return false
+	}
+	return wl.Status.Cluster == nil
+}
+
+// add adds a new Controller to mgr with r as the reconcile.Reconciler
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	// Create a new controller
+	c, err := controller.New(controllerName, mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to Instance
+	err = c.Watch(&source.Kind{Type: &computev1alpha1.Workload{}}, &handler.EnqueueRequestForObject{}, &predicate.Funcs{CreateFunc: CreatePredicate})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// fail - helper function to set fail condition with reason and message
+func (r *Reconciler) fail(instance *computev1alpha1.Workload, reason, msg string) (reconcile.Result, error) {
+	log.Printf("%s: %s", reason, msg)
+	instance.Status.SetCondition(corev1alpha1.NewCondition(corev1alpha1.Failed, reason, msg))
+	return resultRequeue, r.Update(ctx, instance)
+}
+
+// _schedule assigns Workload to a matching cluster. If the workload matches more than one cluster use
+// round-robin to select next cluster
+func (r *Reconciler) _schedule(instance *computev1alpha1.Workload) (reconcile.Result, error) {
+
+	clusters := &computev1alpha1.KubernetesClusterList{}
+
+	err := r.List(context.Background(), client.MatchingLabels(instance.Spec.ClusterSelector), clusters)
+	if err != nil {
+		return resultDone, err
+	}
+
+	if len(clusters.Items) == 0 {
+		return r.fail(instance, errorUnschedulable, "Cannot match to any existing cluster")
+	}
+
+	// round-robin cluster index selection
+	index := int(r.lastClusterIndex % uint64(len(clusters.Items)))
+	cluster := clusters.Items[index]
+	r.lastClusterIndex++
```

Comment:
I wondered what would happen when `r.lastClusterIndex` hits `math.MaxUint64` - it wraps back to zero, potentially skipping over some clusters. The below algorithm avoids that, but I don't think it actually matters in practice given how rarely we're actually going to wrap a `uint64`. 🤓 

```golang
cluster := clusters[r.index]
r.index = (r.index + 1) % uint64(len(clusters))
```

Reply:
I think this should be called "eventually round-robin", since every time we change number of clusters we ran into possibility of either skipping or doubling down on a given cluster index.